Source/DafnyStandardLibraries/src/Std/TargetSpecific/Duration.dfy

```diff
@@ -0,0 +1,405 @@
+      var ParseString := ParseNumericString(s[1..]);
+      var pow := Pow(10, |s| - 1);
+      assume {:axiom} ParseString >= 0;
+      assume {:axiom} digit >= 0;
+      assume {:axiom} pow >= 0;
+      digit * pow + ParseString
+    //digit * (Pow(10, |s| - 1)) + ParseNumericString(s[1..])
+  }
+
+
+  function ParseComponent(text: string, start: int, end: int): uint32
+    requires start >= 0 && end >= 0 && start <= end && end <= |text|
+  {
+    if start >= end then
+      0
+    else
+      var substr := text[start..end];
+      assert |substr| == end - start;  // ADD THIS
+      assert |substr| > 0;  // ADD THIS - since start < end
+      if IsNumeric(substr) then
+        var parsed := ParseNumericString(substr);
+        assume {:axiom} parsed <= 0xFFFFFFFF;
+        parsed as uint32
+      else
+        0
+  }
+
+  function ParseString(text: string): Duration
+    requires |text| >= 2
+    requires text[0..2] == "PT"
+    ensures ParseString(text).Valid()
+  {
+    var len := |text|;
+
+    var hPos := FindCharOrNeg(text, 'H');
+    var mPos := FindCharOrNeg(text, 'M');
+    var sPos := FindCharOrNeg(text, 'S');
+    var dotPos := FindCharOrNeg(text, '.');
+
+    // Extract hour component
+    var hour : uint32 :=
+      if hPos > 2 then ParseComponent(text, 2, hPos) else 0;
+
+    // Extract minute component
+    var minuteStart : int := if hPos > 0 then hPos + 1 else 2;
+    var minute : uint32 :=
+      if mPos > minuteStart then ParseComponent(text, minuteStart, mPos) else 0;
+
+    // Extract second component
+    var secondStart : int := if mPos > 0 then mPos + 1 else 2;
+    var secondEnd : int :=
+      if dotPos > secondStart then dotPos
+      else if sPos > secondStart then sPos
+      else len;
+    var second : uint32 :=
+      if secondEnd > secondStart then ParseComponent(text, secondStart, secondEnd) else 0;
+
+
+    var millisecond : uint32 :=
+      if dotPos > 0 && sPos > dotPos then
+        var raw := ParseComponent(text, dotPos + 1, sPos);
+        if raw < 1000 then raw else 999
+      else
+        0;
+
+
+    var hour_mult := (hour as uint64) * (SECONDS_PER_HOUR as uint64);
+    var minute_mult := (minute as uint64) * (SECONDS_PER_MINUTE as uint64);
+    var totalSeconds_val := hour_mult + minute_mult + (second as uint64);
+    assume {:axiom} totalSeconds_val <= 0xFFFF;
+    var totalSeconds := totalSeconds_val as uint16;
+
+    Duration(totalSeconds, millisecond as uint16)
+  }
+
+
+
+
+  function EpochDifference(epoch1: uint32, epoch2: uint32): Duration
+    ensures EpochDifference(epoch1, epoch2).Valid()
+  {
+    var diff : uint32 := if epoch1 >= epoch2 then (epoch1 - epoch2) as uint32
+                         else (epoch2 - epoch1) as uint32;
+    FromMilliseconds(diff)
+  }
+
+
+/// Test MaxBy with a sequence of durations
+/// This function demonstrates finding the maximum duration in a sequence
+  function MaxByDuration(durations: seq<Duration>): Duration
```

Comment:
Just curious why this feels like a common enough function to include here? I do like having `Seq.MaxBy`, and application code can always just use `MaxBy(durations, Duration.Compare)` directly.